configure.ac

```diff
@@ -49,192 +49,209 @@ AS_IF([test "x$ac_cv_header_iconv_h" = xyes], [
     AC_DEFINE(CONFIG_FRIBIDI, 1, [found fribidi via pkg-config])
-    ])
+])
 
-PKG_CHECK_MODULES([HARFBUZZ], harfbuzz >= 1.2.3, [
+PKG_CHECK_MODULES([HARFBUZZ], [harfbuzz >= 1.2.3], [
     CFLAGS="$CFLAGS $HARFBUZZ_CFLAGS"
     LIBS="$LIBS $HARFBUZZ_LIBS"
     AC_DEFINE(CONFIG_HARFBUZZ, 1, [found harfbuzz via pkg-config])
-    ])
+])
 
 libpng=false
-if test x$enable_test = xyes || test x$enable_compare = xyes; then
-PKG_CHECK_MODULES([LIBPNG], libpng >= 1.2.0, [
-    CFLAGS="$CFLAGS $LIBPNG_CFLAGS"
-    AC_DEFINE(CONFIG_LIBPNG, 1, [found libpng via pkg-config])
-    libpng=true])
-fi
+AS_IF([test "x$enable_test" = xyes || test "x$enable_compare" = xyes], [
+    PKG_CHECK_MODULES([LIBPNG], [libpng >= 1.2.0], [
+        CFLAGS="$CFLAGS $LIBPNG_CFLAGS"
+        AC_DEFINE(CONFIG_LIBPNG, 1, [found libpng via pkg-config])
+        libpng=true
+    ])
+])
 
 ## Check for system font providers
 ### Fontconfig
-if test x$enable_fontconfig != xno; then
-PKG_CHECK_MODULES([FONTCONFIG], fontconfig >= 2.10.92, [
-    CFLAGS="$CFLAGS $FONTCONFIG_CFLAGS"
-    LIBS="$LIBS $FONTCONFIG_LIBS"
-    AC_DEFINE(CONFIG_FONTCONFIG, 1, [found fontconfig via pkg-config])
-	fontconfig=true
-    ], [fontconfig=false])
-fi
+AS_IF([test "x$enable_fontconfig" != xno], [
+    PKG_CHECK_MODULES([FONTCONFIG], [fontconfig >= 2.10.92], [
+        CFLAGS="$CFLAGS $FONTCONFIG_CFLAGS"
+        LIBS="$LIBS $FONTCONFIG_LIBS"
+        AC_DEFINE(CONFIG_FONTCONFIG, 1, [found fontconfig via pkg-config])
+        fontconfig=true
+    ],[
+        fontconfig=false
+    ])
+])
 
 ### Coretext
-if test x$enable_coretext != xno; then
-# Linking to CoreText directly only works from Mountain Lion and iOS.
-# In earlier OS X releases CoreText was part of the ApplicationServices
-# umbrella framework.
-AC_MSG_CHECKING([for CORETEXT])
-AC_COMPILE_IFELSE([
-  AC_LANG_PROGRAM(
-    [[#include <ApplicationServices/ApplicationServices.h>]],
-    [[CTFontDescriptorCopyAttribute(NULL, kCTFontURLAttribute);]])
-  ], [
-    LIBS="$LIBS -framework ApplicationServices -framework CoreFoundation"
-    AC_DEFINE(CONFIG_CORETEXT, 1, [found CoreText in ApplicationServices framework])
-    coretext=true
-    AC_MSG_RESULT([yes])
-  ], [
+AS_IF([test "x$enable_coretext" != xno], [
+    # Linking to CoreText directly only works from Mountain Lion and iOS.
+    # In earlier OS X releases CoreText was part of the ApplicationServices
+    # umbrella framework.
+    AC_MSG_CHECKING([for CORETEXT])
     AC_COMPILE_IFELSE([
-      AC_LANG_PROGRAM(
-        [[#include <CoreText/CoreText.h>]],
-        [[CTFontDescriptorCopyAttribute(NULL, kCTFontURLAttribute);]])
-      ], [
-        LIBS="$LIBS -framework CoreText -framework CoreFoundation"
-        AC_DEFINE(CONFIG_CORETEXT, 1, [found CoreText framework])
+        AC_LANG_PROGRAM( dnl#First test for legacy include
+            [[#include <ApplicationServices/ApplicationServices.h>]]
+            ,[[CTFontDescriptorCopyAttribute(NULL, kCTFontURLAttribute);]]
```

Comment:
I agree that comma at the beginning looks a bit weird. However, this is not a natural text, so "basic typography rules" won't necessarily lead to the best result functionality-wise.
`[]`-M4-Quotes already provide a strong visual separation of arguments, which makes it easy to miss a comma in-between (intended to be separate) arguments, or commas at eol; M4 will happily merge two adjacent quoted sections into one argument.
Furthermore autoconf's M4-macros (in this case unfortunately) have most parameters optional, so such a mistake might not be immediately noticed. As was the case with `LARGE_TILES`, which was broken for ca seven years.
Putting the comma at the beginning makes it stick out like a sore thumb and by extension misplaced commas will be very hard to miss, preventing errors like with `LARGE_TILES`.

Imo the gains in error-prevention outweigh the aesthetic misstep of putting commas at the beginning.

Reply:
Broken case of `LARGE_TILES` was due to the dangling closing bracket (i. e. broken typography rule). And I think that forceful redirection of attention due to a weird comma placement is not helping in error-prevention in general. Maybe it indeed improve visibility of the missing comma in that particular place at the cost of readability, but attention has redirected from somewhere and error in that place would be less visible.

Reply:
As you said, style decisions are partly subjective. I think it's valid to make the logical separator `,` stand out, given that it can be easily missed next to quoting brackets. Due to many parameters being optional and superfluous params being ignored, you also can't rely on M4 to tell you when a `,` has been misplaced *(and if you're unlucky the resulting `configure` is also valid, but does different things than intended)*.

However, the style should be acceptable for all involved. If I were the only one who finds it useful, there would be no point in using this style. So: @astiob @rcombs :
What do you think? Should we use the leading commas to aid in preventing errors due to misplaced commas even if it looks a bit unusual, or do you prefer to use more common commas at eol relying on misplacements being found in review anyway?

Reply:
Personally, I find commas on the same line more ~~pleasant~~ usual, but from what I understand that’s part of the point and people get used to leading commas quickly.

Generally, this same discussion could be had about any other list with separators within our code base, and there are indeed code styles out there that lead their lines with commas for all the same reasons. My impression from @TheOneric’s explanations is that this file is special because a missing comma is harder to spot in M4 than in C because it produces no compilation error, just silently wrong logic. Additionally, I imagine people generally have a lower understanding of M4/Autoconf than of C, so someone who does spot a missing line-trailing comma may have doubts as to whether it’s a bug, while a missing line-leading comma in a file full of line-leading commas might seem more obviously like an error that should be reported/investigated/fixed.

Honestly, I’m overall kinda neutral on this. I understand both sides of the argument and I can live with either decision.

One observation I have is that the first line does not have the line-leading comma, which itself is jarring. On the other hand if we stick with line-trailing commas, does M4 freely allow a trailing comma after the last argument like the one we saw in that Windows check? In languages/contexts that allow them, I prefer to put a comma after every element (including the last one) in multi-line expressions to make them uniform and help avoid missing-comma bugs & keep diffs cleaner when reordering or adding lines/elements.

Reply:
Clarification edit:
> In languages/contexts that allow them, I prefer to put a comma after every element (including the last one) in many-line expressions to make them uniform and help avoid missing-comma bugs & keep diffs cleaner when reordering or adding lines/elements, and to place the closing parenthesis/bracket on its own line.

Reply:
> does M4 freely allow a trailing comma after the last argument like the one we saw in that Windows check?

Mostly yes, for M4, but not quite for Autoconf.
As for M4 itself, `“excess arguments are ignored.”`, but for **builtins**, unless surpressed, a warning about too many / too few args is issued. In (recent) Autoconf this warning apparently becomes fatal (try eg: `m4_index([abc],,)`).
User defined macros would need to check the arg count themselves; after testing a few it seems like Autoconf macros don't (currently) do this.

**If** we wanted to use additional trailing commas, we would need to exempt M4-builtins from this and *hope* autoconf doesn't implement a stricter argument policy in the future.
Unless I missed Autoconf's promise to always ignore superfluous arguments in non-m4-builtins, I'd prefer to not rely on that.